File: audit/python-compatibility/run-cpython.py
"""Bounded, isolated audit oracle. Never imported by the learning app."""
import contextlib, io, json, os, pathlib, sys, traceback, unicodedata
request = json.load(sys.stdin)
root = pathlib.Path(request['directory']).resolve()
root.mkdir(parents=True, exist_ok=True)
os.chdir(root)
sys.path.insert(0, str(root))
for file in request.get('files', []):
    path = pathlib.Path(file['name'])
    if path.name != file['name']:
        raise RuntimeError('Audit fixture must be a flat relative name')
    path.write_bytes(file['content'].encode(file.get('encoding', 'utf-8')))
source = request['code']
stdout, stderr = io.StringIO(), io.StringIO()
sys.stdin = io.StringIO('\n'.join(request.get('input', [])) + ('\n' if request.get('input') else ''))
result = {'stdout': '', 'errorType': None, 'message': None, 'file': None, 'line': None, 'column': None, 'completion': 'complete'}
with contextlib.redirect_stdout(stdout), contextlib.redirect_stderr(stderr):
    try:
        exec(compile(source, 'main.py', 'exec'), {'__name__': '__main__', '__file__': str(root / 'main.py')})
    except BaseException as error:
        result.update(errorType=type(error).__name__, message=str(error), completion='error')
        if isinstance(error, SyntaxError):
            result.update(file=pathlib.Path(error.filename or 'main.py').name, line=error.lineno, column=error.offset)
        else:
            frames = traceback.extract_tb(error.__traceback__)
            relevant = [f for f in frames if f.filename == 'main.py' or pathlib.Path(f.filename).parent == root]
            if relevant:
                f = relevant[-1]
                result.update(file=pathlib.Path(f.filename).name, line=f.lineno, column=(f.colno + 1 if f.colno is not None else None))
result['stdout'] = stdout.getvalue()
result['files'] = {}
for path in root.iterdir():
    if path.is_file() and path.suffix in ('.txt', '.csv'):
        raw = path.read_bytes()
        encoding = next((f.get('encoding', 'utf-8') for f in request.get('files', []) if f['name'] == path.name), 'utf-8')
        try: result['files'][path.name] = raw.decode(encoding)
        except UnicodeError: result['files'][path.name] = {'hex': raw.hex(), 'decodeError': True}
result['environment'] = {'python': sys.version, 'executable': sys.executable, 'unicode': unicodedata.unidata_version, 'recursionLimit': sys.getrecursionlimit()}
print(json.dumps(result, ensure_ascii=True))

File: audit/python-compatibility/run-cpython.mjs
import { spawnSync } from 'node:child_process';
import { resolve } from 'node:path';
export const python = process.env.AUDIT_PYTHON;
if (!python) throw Error('Set AUDIT_PYTHON to the CPython 3.12.14 executable before running this audit.');
const version = spawnSync(python, ['-I', '-c', 'import platform; print(platform.python_version())'], { encoding: 'utf8' });
if (version.status !== 0 || version.stdout.trim() !== '3.12.14') throw Error('This audit requires CPython 3.12.14; other versions must not be mixed.');
export function runCPython(c) {
  if (c.vmOnly) return { skipped: true, reason: c.vmOnly, stdout: '', errorType: null, completion: 'not-run' };
  const r = spawnSync(python, ['-I', '-X', 'utf8', resolve('audit/python-compatibility/run-cpython.py')], { input: JSON.stringify({ ...c, directory: resolve('artifacts/python-compatibility-work', c.id) }), encoding: 'utf8', timeout: 5000, maxBuffer: 2_000_000, env: { ...process.env, PYTHONHASHSEED: '0' } });
  if (r.error || r.status !== 0) return { stdout: r.stdout ?? '', errorType: r.error?.code ?? 'OracleProcessError', message: r.stderr, completion: 'harness-error' };
  return JSON.parse(r.stdout);
}
